### ISSUE TITLE
AtomicConv Speed Optimizations

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#1412]
I ran the `AtomicConvModel` against test systems of different sizes to measure how long a single gradient descent step took.

- 50 ligand atoms, 50 protein atoms: 14 seconds
- 100 ligand atoms, 100 protein atoms: 19 seconds
- 200 ligand atoms, 200 protein atoms: 36 seconds
- 400 ligand atoms, 400 protein atoms: 109 seconds 
- 800 ligand atoms, 800 protein atoms: 552 seconds
- 1000 ligand atoms, 1200 protein atoms: 1036 seconds

These times are quite slow (I'm on a CPU machine though, so there may be some GPU speedup, but relative scales are probably similar). Even training on protein-ligand binding pockets (#1404), model training would be quite slow. In addition, it wouldn't be feasible to use an atomic conv for protein-protein or protein-dna systems (both of which datasets are available in PDBBind for example).

Are there any speed-ups or architectural tweaks we can make to the atomic conv layers to speed up usage? Perhaps variants of the atomic conv layer that don't generate tensors that large internally. I think getting #1404 implemented will be critical for these improvements since we'll need to benchmark to make sure tweaks don't damage learning performance

**Comment by rbharath** [comment]
I'm working on running some of these numbers against a GPU machine, and one thing I notice is that GPU utilization is pretty low. At times spikes up to 50%, but mostly hovering between 0-10%. Perhaps preprocessing is the heavy cost?

**Comment by peastman** [comment]
I'll take a look and see if I can find anything.  Can you post or email the files to run your benchmark?

**Comment by rbharath** [comment]
Great, thanks! I used one of the atomic conv unit tests `test_atomic_conv_variable` and changed the variables `frag1_num_atoms` and `frag2_num_atoms` to alter the size of the atoms the test ran on.

```
nosetests -v deepchem/models/tensorgraph/tests/test_atomic_conv.py:TestAtomicConv.test_atomic_conv_variable --nologcapture --nocapture
```

**Comment by rbharath** [comment]
This uses code from the latest merged PR #1411

**Comment by peastman** [comment]
If you can get #1392 working (support eager mode), that would make it a lot easier to profile.

**Comment by peastman** [comment]
It looks like `distance_tensor()` is now a major bottleneck.  It's written in a very inefficient way with a bunch of stack and unstack operations.  It starts with the atom positions and neighbor list, each in a single tensor.  Then it splits them into one tensor per atom, then loops over each atom and computes the distances for each one, in the process doing lots more splits and merges, and then stacks them all back together again.

**Comment by prateethvnayak** [comment]
Any updates on this !? I'm experiencing very poor GPU utilization even on training just 900 smiles string using the `GraphConv` or `MPNN` model for determining solubility.

**Comment by peastman** [comment]
The fix was in #1413.  There may still be opportunities for optimization, but it's a lot faster now than it used to be.  Also, this issue is about atomic convolutions.  GraphConvModel and MPNNModel are unrelated to it.

**Comment by prateethvnayak** [comment]
Sorry. 

> The fix was in #1413. There may still be opportunities for optimization, but it's a lot faster now than it used to be. Also, this issue is about atomic convolutions. GraphConvModel and MPNNModel are unrelated to it.

Sorry, will move to a new issue or related one. Will be deleting the comment to prevent any confusion.
